### DISCUSSION TITLE
How do I control what the sizeof(uint32_t) is?

### DISCUSSION THREAD

**Discussion by dhylands** [microsoft/vscode-cpptools#11778] [General]
If I create this file:
```
#include <cassert>
#include <cstddef>
#include <cstdint>

static constexpr size_t FLOAT_SIZE = sizeof(float);
static constexpr size_t UINT32_SIZE = sizeof(uint32_t);
static constexpr size_t UINT_SIZE = sizeof(unsigned int);
static constexpr size_t UINT_LONG_SIZE = sizeof(unsigned long int);


static_assert(sizeof(uint32_t) == sizeof(float));
```
then the static_assert appears with red squiggles under it. If I hover over the `FLOAT_SIZE` variable it shows me that it has a value of `4`.
If I hover over the `uint32_t` variable then it shows me that it has a value of `8`. 
Hovering over `UINT_SIZE` shows `4` which should be `2` for the compiler I'm using.
Hovering over `UINT_LONG_SIZE` shows `8` which should be `4` for the compiler I'm using.

I'm using a MSP430 compiler which has 16-bit ints, so `sizeof(unsigned int)` is `2` and `sizeof(unsigned long int)` is `4`.

This particular codebase is being compiled targeting the MSP430 GCC compiler. I already have:
```
 "C_Cpp.default.compilerPath": "/home/dhylands/ti/msp430-gcc/bin/msp430-elf-g++",
 ```
 set in my settings.json file.  I also have 
```
        "/home/dhylands/ti/msp430-gcc/include",
        "/home/dhylands/ti/msp430-gcc/msp430-elf/include",
```
in my  "C_Cpp.default.includePath" which is why it's detecting that `uint32_t` is the type `unsigned long`.

Is there someway to tell the extension that sizeof(int) is 2 rather than 4? (and that `sizeof(unsigned long)` is `4`?)

**Comment by Colengms** [comment]
Hi @dhylands .  It shouldn't be necessary to specify a compiler's system include paths.  If you're configuring the C/C++ Extension with a `c_cpp_properties.json` or the equivalent `C_Cpp.default` settings, you will want to specify the compiler you are using in the `compilerPath` field, as well as any arguments that might influence the behavior of the compiler, in the `compilerArgs` field.  The C/C++ Extension will invoke the compiler with these arguments to solicit system include paths and system defines from the compiler.  If using a custom configuration provider, or configured with a `compile_commands.json`, those should already provide the compiler and arguments to the C/C++ Extension.

However, the issue you're encountering may (also) be related to the C/C++ Extension currently only supporting a fixed set of `IntelliSenseMode`'s, which dictate certain internal details related to compiler/platform/architecture, such as the sizes of certain standard types.  We're working to make more of this detectable by the compiler query, and to support more target platforms and architectures.  That is tracked by: https://github.com/microsoft/vscode-cpptools/issues/7355   In the meanwhile, if you're not able to resolve your issue by specifying a compiler path, it may be possible to work around the issue by modifying an internal JSON files under the extensions `Extension\bin` path, to repurpose one of the existing `IntelliSenseMode`s to match your specific target platform and architectures.

**Comment by dhylands** [comment]
My compilerPath is set like this: `"C_Cpp.default.compilerPath": "/home/dhylands/ti/msp430-gcc/bin/msp430-elf-g++"`
and if I run:
```
/home/dhylands/ti/msp430-gcc/bin/msp430-elf-g++ -x c++ -E -dM /dev/null | grep -i SIZEOF
```
then I get this output:
```
#define __SIZEOF_INT__ 2
#define __SIZEOF_POINTER__ 2
#define __SIZEOF_LONG__ 4
#define __SIZEOF_LONG_DOUBLE__ 8
#define __SIZEOF_SIZE_T__ 2
#define __SIZEOF_WINT_T__ 2
#define __SIZEOF_PTRDIFF_T__ 2
#define __SIZEOF_FLOAT__ 4
#define __SIZEOF_SHORT__ 2
#define __SIZEOF_INT20__ 4
#define __SIZEOF_WCHAR_T__ 4
#define __SIZEOF_DOUBLE__ 8
#define __SIZEOF_LONG_LONG__ 8
```
If my source file contains:
```
    int x = __SIZEOF_INT__;
    int y = sizeof(int);
```
and I hover over the `__SIZEOF_INT__` then it shows a value of 2. But if I hover over the `sizeof(int)` it shows `(int)4`.
